src/utils.jl

```diff
@@ -156,8 +156,5 @@ function subHv(p::AbstractArray{T}, x::AbstractArray{T}, g::AbstractArray{T}, Hv
     return f, g
 end
 
-# THresholding
-soft_thresholding(x::AbstractArray{Complex{T}}, λ::T) where {T} = exp.(angle.(x)im) .* max.(abs.(x) .- convert(T, λ), T(0))
-soft_thresholding(x::AbstractArray{Complex{T}}, λ::Array{T}) where {T} = exp.(angle.(x)im) .* max.(abs.(x) .- convert(Array{T}, λ), T(0))
-soft_thresholding(x::AbstractArray{T}, λ::T) where {T} = sign.(x) .* max.(abs.(x) .- convert(T, λ), T(0))
-soft_thresholding(x::AbstractArray{T}, λ::Array{T}) where {T} = sign.(x) .* max.(abs.(x) .- convert(Array{T}, λ), T(0))
+# Thresholding
+soft_thresholding(x::Union{AbstractArray{Complex{T}},AbstractArray{T}}, λ::Union{T, Array{T}}) where {T} = x .* max.(T(1) .- λ ./ abs.(x), T(0))
```

Comment:
I personally think this scaling way is cleaner than dispatching to complex/real case. Also performance-wise it is better since `abs` is faster than `angle`.

Reply:
It's also very unstable which is wrong. Assuming there is no small/zeros in x for a sparisty promoting algorithm  is not a good idea at all

Reply:
For small/zero in `x`, this term `λ ./ abs.(x)` will be infinity. So overall it will end up to be `0`. So I don't think that's a problem?

Reply:
Dividing by zero.is generally  a bad idea (and definitely  won't gain you performance).

- if you want performance dispatch will always win
- if you want performance you would have to look at `ifelse`
- if the implementation  doesn't reflect the math it needs proper documentation

Reply:
My point is not on dispatch but on the usage of `angle` as it is much slower than `abs`. But yea seems like we would have to do it otherwise there is division by 0 issue.